### DISCUSSION TITLE
[JS201]物件導向課程影片提問

### DISCUSSION THREAD

**Discussion by s103071049** [Lidemy/forum#171] [其他問題]
下面代碼來自課程影片介紹 es5 以前可以如何實作 class 的方式

`console.log(d.sayHello === b.sayHello)` 透過這樣的操作，我們知道這兩個是不同的 function

我的疑惑在於為甚麼今天，我這樣寫`console.log(d.sayHello( ) === b.sayHello( ))` 會印出 true，我明明對 `d.sayHello() ` 與 `b.sayHello()` 操作得到的是 abc 跟 xxx

用 console.log 分別印出 `console.log(d.sayHello())` 與 `console.log(b.sayHello())` 都出現 undefined

問題一、 `console.log(d.sayHello())` 與 `console.log(b.sayHello())` 都出現 undefined 是因為這邊相當於是印出呼叫函式完成的結果，sayHello return 的是 undefined，所以都印出了 undefined，對嗎 ?

問題二、`d.sayHello() ` 與 `b.sayHello()` 操作得到的是 abc 跟 xxx，這邊也是 call function，可是為甚麼回傳的方式和問題一不同 ?

```js
function Dog(name) {
  var myName = name
  return {
    getName: function() {
      return myName
    },
    sayHello: function() {
      console.log(myName)
    }
  }
}
var d = Dog('abc')
d.sayHello() // abc

var b = Dog('xxx')
b.sayHello() // xxx
```

**Comment by EshauHuang** [comment]
>我的疑惑在於為甚麼今天，我這樣寫console.log(d.sayHello( ) === b.sayHello( )) 會印出 true，我明明對 d.sayHello() 與 b.sayHello() 操作得到的是 abc 跟 xxx

因為兩個回傳值都是 `undefined` 因此會是 true
```
console.log(d.sayHello( ) === b.sayHello( ))  //console.log(undefined === undefined)，true
```

> 問題一、 console.log(d.sayHello()) 與 console.log(b.sayHello()) 都出現 undefined 是因為這邊相當於是印出呼叫函式完成的結果，sayHello return 的是 undefined，所以都印出了 undefined，對嗎 ?

因為 `d.sayHello()` 與 `b.sayHello()` 都是在呼叫 function，而你的 function 內並沒有設定回傳值，因此回傳預設值 undefined，參考[這篇](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters#description)
> In JavaScript, function parameters default to undefined.

在 JavaScript 中，函式的參數預設值都為 undefined，舉個例子的話就像
```
function test() {
  console.log('test')
}
console.log(test())  //印出 test undefined
```
不過我猜第一題你應該是想用 `console.log(d.getName())` 及 `console.log(b.getName())` 吧？這樣就會跟第二題所說的值相同。
>問題二、d.sayHello() 與 b.sayHello() 操作得到的是 abc 跟 xxx，這邊也是 call function，可是為甚麼回傳的方式和問題一不同 ?

`console.log` 是印出值，`return` 是回傳值，要印出「回傳值」可以用 `console.log`
```
function test() {
  return 'test'
}
console.log(test())  //test
```

**Reply by s103071049** [reply]
感謝回覆，被解答到了，謝謝!!!

**Comment by aszx87410** [comment]
可以回去複習一下 JS101 的部分，裡面有提到 console.log 跟 return 差在哪邊，你的問題應該就只是這兩個搞混了而已，比較細節的部分樓上的同學回答得很清楚了～

**Reply by s103071049** [reply]
老師謝謝XD
